processor/src/errors.rs

```diff
@@ -174,6 +186,97 @@ pub enum Ext2InttError {
     UninitializedMemoryAddress(u32),
 }
 
+// ERROR CONTEXT
+// ===============================================================================================
+
+#[derive(Debug)]
+pub struct ErrorContext<'a, N: MastNodeExt>(Option<ErrorContextImpl<'a, N>>);
+
+impl<'a, N: MastNodeExt> ErrorContext<'a, N> {
+    pub fn new(
+        mast_forest: &'a MastForest,
+        node: &'a N,
+        source_manager: Arc<dyn SourceManager>,
+    ) -> Self {
+        Self(Some(ErrorContextImpl::new(mast_forest, node, source_manager)))
+    }
+
+    pub fn new_with_op_idx(
+        mast_forest: &'a MastForest,
+        node: &'a N,
+        source_manager: Arc<dyn SourceManager>,
+        op_idx: usize,
+    ) -> Self {
+        Self(Some(ErrorContextImpl::new_with_op_idx(
+            mast_forest,
+            node,
+            source_manager,
+            op_idx,
+        )))
+    }
+
+    pub fn none() -> Self {
+        Self(None)
+    }
+
+    pub fn label_and_source_file(&self) -> (Option<SourceSpan>, Option<Arc<SourceFile>>) {
+        self.0.as_ref().map_or((None, None), |ctx| ctx.label_and_source_file())
+    }
+}
+
+impl Default for ErrorContext<'_, BasicBlockNode> {
+    fn default() -> Self {
+        Self::none()
+    }
+}
+
+#[derive(Debug)]
+struct ErrorContextImpl<'a, N: MastNodeExt> {
+    mast_forest: &'a MastForest,
+    node: &'a N,
+    source_manager: Arc<dyn SourceManager>,
+    op_idx: Option<usize>,
+}
```

Comment:
Why could we not just use `MastNode` here (instead of `MastNodeExt`)? We could probably add `MastNode::get_assembly_op()` function which would internally just call the trait implementation.

Reply:
Not really, because in almost all cases where we construct an `ErrorContext`, we have access to a reference of an unwrapped `MastNode` - e.g. `&CallNode` or `&BasicBlockNode`. To use `MastNode` in here, we'd need an owned `MastNode`, and at each call site, we'd do e.g. `MastNode::CallNode(call_node_ref.clone())`, which is obviously undesirable.

Reply:
Ah - makes sense. The main thing I dislike here is that we have to have this generic parameter which we need to specify every time we pass `ErrorContext` as a parameter. But not sure if there is a good way to get rid of it.

Reply:
> But not sure if there is a good way to get rid of it.

Not that I know of 🤷‍♂️